src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/VisitResourceIntTest.kt

```diff
@@ -174,9 +176,69 @@ class VisitResourceIntTest : IntegrationTestBase() {
 
         val balanceAdjustment = offenderVisitBalanceAdjustmentRepository.findAll()
 
-        assertThat(balanceAdjustment).extracting("offenderBooking.bookingId", "remainingVisitOrders").containsExactly(
-          Tuple.tuple(offenderBookingId, -1),
+        assertThat(balanceAdjustment).extracting("offenderBooking.bookingId", "remainingPrivilegedVisitOrders")
+          .containsExactly(
+            Tuple.tuple(offenderBookingId, -1),
+          )
+        offenderVisitBalanceAdjustmentRepository.deleteAll()
+      }
+    }
+  }
+
+  @DisplayName("Cancel")
+  @Nested
+  inner class CancelVisitRequest {
+    @Test
+    fun `cancel visit success`() {
+      val createResponse = webTestClient.post().uri("/prisoners/$offenderNo/visits")
+        .headers(setAuthorisation(roles = listOf("ROLE_UPDATE_NOMIS")))
+        .contentType(MediaType.APPLICATION_JSON)
+        .body(BodyInserters.fromValue(createVisitRequest))
+        .exchange()
+        .expectStatus().isCreated
+        .expectBody(CreateVisitResponse::class.java)
+        .returnResult().responseBody
+      val visitId = createResponse?.visitId!!
+
+      webTestClient.put().uri("/prisoners/$offenderNo/visits/$visitId/cancel")
+        .headers(setAuthorisation(roles = listOf("ROLE_UPDATE_NOMIS")))
+        .contentType(MediaType.APPLICATION_JSON)
+        .body(
+          BodyInserters.fromValue(
+            """{
+            "outcome" : "VISCANC"
+          }"""
+          )
         )
+        .exchange()
+        .expectStatus().isOk
+        .expectBody().isEmpty
+
+      // Spot check that the database has been updated correctly.
+      TransactionTemplate(transactionManager).execute {
+        val visit = visitRepository.findById(visitId).orElseThrow()
+        assertThat(visit.visitStatus?.code).isEqualTo("CANC")
+        assertThat(visit.offenderBooking?.bookingId).isEqualTo(offenderBookingId)
+        assertThat(visit.visitors).extracting("eventOutcome.code", "eventStatus.code", "outcomeReason.code")
+          .containsExactly(
+            Tuple.tuple("ABS", "CANC", "VISCANC"),
+            Tuple.tuple("ABS", "CANC", "VISCANC"),
+            Tuple.tuple("ABS", "CANC", "VISCANC"),
+            Tuple.tuple("ABS", "CANC", "VISCANC"),
+          )
+        assertThat(visit.visitOrder?.status?.code).isEqualTo("CANC")
+        assertThat(visit.visitOrder?.outcomeReason?.code).isEqualTo("VISCANC")
+        assertThat(visit.visitOrder?.expiryDate).isEqualTo(LocalDate.now())
+
+        val balanceAdjustments = offenderVisitBalanceAdjustmentRepository.findAll()
+
+        assertThat(balanceAdjustments).extracting("offenderBooking.bookingId", "remainingPrivilegedVisitOrders")
+          .containsExactly(
+            Tuple.tuple(offenderBookingId, -1),
+            Tuple.tuple(offenderBookingId, 1),
+          )
+        // TODO has to be done separately
+        offenderVisitBalanceAdjustmentRepository.deleteAll()
```

Comment:
Is this something I should look into with the entity builder stuff?

Reply:
👍  That would be nice,  currently they are just left